torch/onnx/_internal/exporter/_torchlib/ops/nn.py

```diff
@@ -122,16 +122,13 @@ def aten_scaled_dot_product_attention_23(
         # NOTE: There was extended discussion on whether the num_heads attributes (q_num_heads/kv_num_heads)
         # should be set as ONNX attributes or inferred from the tensor shape. In ONNX, num_heads is needed
         # for 3D attention inputs (shape: [B, S, N*H]), but not for 4D ([B, N, S, H]), which is the only
-        # input accepted by this exporter. Thus, the attribute is not strictly necessary here, but adding it
-        # may ease future optimization or conversion to 3D formats (e.g., GQA ops)
+        # input accepted by this exporter.
         Y, _, _, _ = op23.Attention(
             query,
             key,
             value,
             attn_mask=attn_mask,
             scale=scale,
-            q_num_heads=query.shape[-3],
```

Comment:
Could you explain? Thanks

cc @yuanyao-nv

Reply:
The comment in the code mentions that. I would say if onnx does not need it, it is better not to add them. Right now, onnx implementation does something different when these attributes are populated or not populated for 4D inputs when the specifications seem to say otherwise. But it is not very clear to me. Right now, I'm just trying to make the code work.

Reply:
Just to add a quick summary here (of a long offline discussion): the ONNX function-definition (for Attention) assumes that these attributes are specified IF and ONLY IF the input is the 3D case. But the ONNX op documentation does not make this clear.  So, the simplest solution for now is to drop these attributes here. Other tools/optimizers will need to use the shape-information to figure it out if they need it.